store/auth.go

```diff
@@ -80,45 +82,107 @@ func loginCaveatID(m *macaroon.Macaroon) (string, error) {
 	return caveatID, nil
 }
 
+// retryRequestDecodeJSON calls retryRequest and decodes the response into either success or failure.
+func retryRequestDecodeJSON(endpoint string, headers map[string]string, reader io.Reader, success interface{}, failure interface{}) (resp *http.Response, err error) {
+	return retryRequest(endpoint, headers, reader, func(ok bool, resp *http.Response) error {
+		result := success
+		if !ok {
+			result = failure
+		}
+		if result != nil {
+			return json.NewDecoder(resp.Body).Decode(result)
+		}
+		return nil
+	})
+}
+
+// retryRequest calls doRequest and decodes the response in a retry loop.
+func retryRequest(endpoint string, headers map[string]string, reader io.Reader, decode func(ok bool, resp *http.Response) error) (resp *http.Response, err error) {
```

Comment:
don't we an issue with io.Reader about trying to read it again?  at least in store.go we ended up using []byte instead?